Add stm32 DMA driver support

cybertale opened the pull request:
This PR adds driver support for DMA on f0/f1/f2/f3/f4/l0/l4 series stm32.
This should have been working on the f7 series SoCs, but there are some weird bugs inside the code that I cannot resolve.
There are two kinds of IP blocks are used across these stm32, one is the one that has been used on F2/F4/F7 series, and the other one is the one that has been used on F0/F1/F3/L0/L4 series.

Memory to memory transfer is only supported on the second DMA on F2/F4/F7 with 'st,mem2mem' to be declared in dts.

This driver depends on k_malloc to allocate memory for stream instances, so CONFIG_HEAP_MEM_POOL_SIZE must be big enough to hold them.

Common parts of the driver are in dma_stm32.c and SoC related parts are implemented in dma_stm32_v*.c.

This driver has been tested on multiple nucleo boards, including NUCLEO_F091RC/F103RB/F207ZG/F302R8/F401RE/L073RZ/L476RG with the loop_transfer and chan_blen_transfer test cases.

Fixes #9193
This is the driver part divided from #13364 .
This is based on #19711 .

avisconti commented on drivers/dma/dma_stm32.c:
```diff
@@ -0,0 +1,617 @@
+			config->dma_slot = 0;
+		}
+	}
+	DMA_InitStruct.Channel = table_ll_channel[config->dma_slot];
```
table_ll_channel[] is declared as static in another file. So, it does not compile correctly.

avisconti replied:
@cybertale Even if it is strange that does not build, can you take a look at this issue?

avisconti replied:
Oh, I noticed now that the issue was found by CI as well...

cybertale replied:
Yes, I'm just fixing this, come back to check the new driver about 1 hour later. Should be fixed then :smile:

cybertale commented:
@avisconti Problems should be all fixed. Give it a try when you got time. Also have a look on the macro 
we talked about on this commit https://github.com/zephyrproject-rtos/zephyr/pull/19711/commits/3dc047eaa208c66207fd2d24573e09b727bb1b83

avisconti commented:
> @avisconti Problems should be all fixed. Give it a try when you got time. Also have a look on the macro
> we talked about on this commit [3dc047e](https://github.com/zephyrproject-rtos/zephyr/commit/3dc047eaa208c66207fd2d24573e09b727bb1b83)

I was expecting it to work straight away, but it is not. (sigh!)
I will take a deeper look tomorrow. I don't know... everything seems fine.

cybertale commented:
> > @avisconti Problems should be all fixed. Give it a try when you got time. Also have a look on the macro
> > we talked about on this commit [3dc047e](https://github.com/zephyrproject-rtos/zephyr/commit/3dc047eaa208c66207fd2d24573e09b727bb1b83)
> 
> I was expecting it to work straight away, but it is not. (sigh!)
> I will take a deeper look tomorrow. I don't know... everything seems fine.

Have you checked the i2s driver with only #19711 ? I think this could help to identify which PR the problem lie in.

avisconti commented:
> > > @avisconti Problems should be all fixed. Give it a try when you got time. Also have a look on the macro
> > > we talked about on this commit [3dc047e](https://github.com/zephyrproject-rtos/zephyr/commit/3dc047eaa208c66207fd2d24573e09b727bb1b83)
> > 
> > 
> > I was expecting it to work straight away, but it is not. (sigh!)
> > I will take a deeper look tomorrow. I don't know... everything seems fine.
> 
> Have you checked the i2s driver with only #19711 ? I think this could help to identify which PR the problem lie in.

I found the issue. It is not in your commits, but in the sample itself. I had already found it last week, but for some reason, since the changes are a lot, yesterday I have used an old sample version.

I'm going to submit a patch for that.

BTW, I think that maybe we should squash i2s changes in the same commit as new dma. And maybe also this commit that I'm going to push in my github.

avisconti commented:
@cybertale 
You can cherry-pick the a1f5802f commit in my branch:
https://github.com/avisconti/zephyr/pull/new/test-add-dma-stm32

Then, we may decide whether to squash it or not with dma changes

erwango commented:
@avisconti, if possible (so unless there are dependencies) it is better to split and merge in small reviewable PR.

cybertale commented:
@erwango So @avisconti create another PR and these three PRs get to be reviewed and merged together, right? 
By the way, does anyone of you got an idea of what's the problem that the CI is complaining about?
Hard to relate the problem to my changes...

erwango commented:
> By the way, does anyone of you got an idea of what's the problem that the CI is complaining about?

qemu targets tend to fails randomly in CI. You just need to rerun.

avisconti commented:
> @erwango So @avisconti create another PR and these three PRs get to be reviewed and merged together, right?

@cybertale @erwango 
You are talking about commits, correct?

Anyway, I would merge in a single commit all changes that are atomic to the new driver, otherwise some things may not work if one commit requires another commit. For example in eeb0e28 you modify the src/dst data size to '2', but this requires 5cbe6ca otherwise the size is not correct. Do you agree? So I will squash this two together.

My commit a1f5802 is a fix that may be left in a separate commit as it should be there anyway, even in the old driver. I can take care of it right now.

cybertale commented:
@erwango What else do we need to get this merged?

erwango commented:
At least, I'd like @avisconti to +1 first.

avisconti commented:
> @cybertale At least, I'd like @avisconti to +1 first.

I have tested few days ago and it was working fine.
This seems to be a separate set of patches, but seems fine to me.

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -0,0 +1,617 @@
+#define IRQ_INIT(dma, chan)                                             \
+do {									\
+	if (!irq_is_enabled(DT_DMA_##dma##_IRQ_##chan)) {		\
+		irq_connect_dynamic(DT_DMA_##dma##_IRQ_##chan,		\
```
Can you remind me why this is needed?

cybertale replied:
In some series of SoCs, DMA irq lines are shared across some channels. Static declaration will raise an error when a irq line has been connected to the driver, so it must be dynamicly checked and connected.

erwango reviewed (changes requested):
> drivers: i2s: code refactor 

Please extract as many i2s changes as possible, and split into more commit since some are fixes and some are refactor.

> scripts: dts: add support for parsing stm32 dma consumer cells

Please update commit title since script is no more impacted


> drivers: dma: add generic driver support for all stm32
This should have been working on the f7 series SoCs, but there
are some weird bugs inside the code that I cannot resolve.

Still valid ?

avisconti commented:
> > drivers: i2s: code refactor
> 
> Please extract as many i2s changes as possible, and split into more commit since some are fixes and some are refactor.

@erwango and @cybertale 
Plus, I would also squash together the i2s changes (not the i2s refactor) with dma changes.
Does it make sense to you?
They are interdependent.

erwango commented:
> Does it make sense to you?

I'd like to split at maximum to be honest.
Any way to have a I2S PR that includes the DMA one (since I2S can live with this change temporaly)?

@galak, any preference?

avisconti commented:
> > Does it make sense to you?
> 
> I'd like to split at maximum to be honest.
> Any way to have a I2S PR that includes the DMA one (since I2S can live with this change temporaly)?
> 
> @galak, any preference?

Actually I don't think it is a matter of preference.
See my previous comment:

"Anyway, I would merge in a single commit all changes that are atomic to the new driver, otherwise some things may not work if one commit requires another commit. For example in eeb0e28 you modify the src/dst data size to '2', but this requires 5cbe6ca otherwise the size is not correct. Do you agree? So I will squash this two together."

avisconti commented:
> @avisconti:
> 
> Isn't it possible to push the STM32 generic DMA driver first while keeping the old one in a first PR,
> and then we update I2S to use the new driver and remove the old one?

Not sure it is possible.
Anyway, don't you think it is confusing?
Maybe @galak can give some recommendations on this specific point.

cybertale commented:
> > drivers: i2s: code refactor
> 
> Please extract as many i2s changes as possible, and split into more commit since some are fixes and some are refactor.
> 

This commit has only one change, that is, using a macro to remove duplicated code in declaration and initialization of the device instances.

> > scripts: dts: add support for parsing stm32 dma consumer cells
> 
> Please update commit title since script is no more impacted
> 
> > drivers: dma: add generic driver support for all stm32
> > This should have been working on the f7 series SoCs, but there
> > are some weird bugs inside the code that I cannot resolve.
> 
> Still valid ?

I'll fix this to the 7 series of SoCs that have passed tests.

cybertale commented on tests/drivers/i2s/i2s_speed/CMakeLists.txt:
```diff
@@ -1,6 +1,13 @@
 
 cmake_minimum_required(VERSION 3.13.1)
+
+if(BOARD STREQUAL "96b_argonkey")
```
But 96b_argonkey.conf is only needed when using this DMA driver, and also it's content depends on our application. In this test 1024 is enough since only DMA driver requires heap memory, while it may be not sufficient for other applications.

cybertale replied:
You're right about this, I misunderstood what you said.

erwango commented:
> > > drivers: i2s: code refactor
> > 
> > 
> > Please extract as many i2s changes as possible, and split into more commit since some are fixes and some are refactor.
> 
> This commit has only one change, that is, using a macro to remove duplicated code in declaration and initialization of the device instances.

My bad, I meant commit "drivers: i2s: apply new dma consumer API to the stm32 i2s driver "

> 
> > > scripts: dts: add support for parsing stm32 dma consumer cells
> > 
> > 
> > Please update commit title since script is no more impacted
> > > drivers: dma: add generic driver support for all stm32
> > > This should have been working on the f7 series SoCs, but there
> > > are some weird bugs inside the code that I cannot resolve.
> > 
> > 
> > Still valid ?
> 
> I'll fix this to the 7 series of SoCs that have passed tests.

cybertale commented:
@erwango OK, I'll see what I can do.

cybertale commented on boards/arm/96b_stm32_sensor_mez/Kconfig.defconfig:
```diff
@@ -53,13 +53,6 @@ config I2S_2
 
 endif # I2S
 
-if DMA
```
For explanation, **DMA** here is selected by **I2S** in Kconfig system now.

cybertale replied:
and DMA_STM32 got selected in soc/arm/st_stm32/common/Kconfig.defconfig.series

cybertale committed:
dt-bindings: add generic dma controller description support

Add a yaml file for help describing generic dma controllers.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

cybertale committed:
dt-bindings: add support for parsing stm32 dma consumer cells

Add support for parsing stm32 dma consumer cells, format of which
follows dma dts format declared in the Linux Kernel for the dma of
stm32:
https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/plain/Bindings/dma/stm32-dma.txt

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
dt-bindings: arm: st: add dts support for DMA of some series of stm32

Add dts support for f0/f1/f2/f3/f4/f7/l0/l4 series STM32.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
drivers: i2s: code refactor

Declaration code for the I2S devices in the driver has too much
duplicate code. This commit uses a help macro to save some work
and some lines of code.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
drivers: i2s: change stm32 i2s driver to use two DMA controllers

The i2s driver assumes the tx channel and rx channel of dma are using
the same dma controller. This commit changes it to be able to use
different dma controllers.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
drivers: i2s: add some missing DMA parameters

src_addr_increment, dst_addr_increment, fifo_threshold and priority
are missing as parameters for configuring DMA. This commit adds them
to the driver.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
drivers: i2s: change the stm32 DMA in the driver to use dts

This commit moves DMA parameters previously hard coded in the driver
to the dts.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

cybertale commented:
# Change Log

1. Split 'drivers: i2s: apply new dma consumer API to the stm32 i2s driver' into more commits.
2. move '96b_argonkey.conf' in i2s tests into boards folder.
3. move DMA dts changes for nucleo_*.dts into tests/drivers/dma/*/boards/nucleo_*.overlay.
4. Some supplement for commit messages.

erwango commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -848,8 +848,8 @@ static struct device *get_dev_from_tx_dma_channel(u32_t dma_channel)
 		.block_count = 1,					\
 		.dma_slot = DT_I2S_##index##_DMA_SLOT_##dir_cap,	\
 		.channel_direction = PERIPHERAL_TO_MEMORY,		\
-		.source_data_size = 1,  /* 16bit default */		\
```
Shouldn't it logically be done after new driver introduction?

cybertale replied:
Yes, this should be after the new driver introduction.

erwango reviewed (changes requested):
Change seems fine a part from commit sequence since we're doing i2s changes related to new driver before driver introduction. This should be reviewed.
Then, ideally I would prefer to split in 2 PR:
- new driver introduction, including dt and samples
- i2s changes along with old driver removal.
But I'm willing to approve as soon as commit order is changed.

Last point is STM32F7 status: can you confirm status provided in new driver commit message is valid ?

Otherwise I'm quite fine with the PR

cybertale commented:
> 
> Last point is STM32F7 status: can you confirm status provided in new driver commit message is valid ?
> 
Yes, I haven't fix the f7 problem, so it's working randomly. Or not working we can treat it as. In the commit message, I added that 'Notice due to some bugs, this is currently not working with f7'. But the driver is completable with f7's DMA architecture. The commit message is updated and valid.

cybertale committed:
dt-bindings: boards: apply dma dts changes to some boards

These boards using i2s in dts now have to declare the usage of dma in
dts, too.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
drivers: dma: add generic driver support for some series of stm32

This commit adds driver support for DMA on f0/f1/f2/f3/f4/l0/l4
series stm32.

Notice due to some bugs, this is currently not working with f7.

There are two kinds of IP blocks are used across these stm32, one is the
one that has been used on F2/F4/F7 series, and the other one is the one
that has been used on F0/F1/F3/L0/L4 series.

Memory to memory transfer is only supported on the second DMA on
F2/F4 with 'st,mem2mem' to be declared in dts.

This driver depends on k_malloc to allocate memory for stream instances,
so CONFIG_HEAP_MEM_POOL_SIZE must be big enough to hold them.

Common parts of the driver are in dma_stm32.c and SoC related parts are
implemented in dma_stm32_v*.c.

This driver has been tested on multiple nucleo boards, including
NUCLEO_F091RC/F103RB/F207ZG/F302R8/F401RE/L073RZ/L476RG with the
loop_transfer and chan_blen_transfer test cases.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale commented:
@erwango How about now?

maxvankessel commented on drivers/dma/dma_stm32.c:
```diff
@@ -0,0 +1,617 @@
+		DMA_InitStruct.FIFOMode = LL_DMA_FIFOMODE_DISABLE;
+	}
+#endif
+	DMA_InitStruct.NbData = config->head_block->block_size;
```
Block size documentation states: `block_size is the number of bytes to be transferred for this block.`
Datasheet on NbData states: `NDT[15:0]: number of data items to transfer (0 up to 65535)` 

So I've tested this, and it happens to be that there is no isr fired when I' m using word sized transfers. 
Only the half transfer interrupt flag is set by the DMA controller.

```suggestion
	DMA_InitStruct.NbData = config->head_block->block_size / config->source_data_size;
```

cybertale replied:
Thanks, I think you're right on this. I've also tested this this morning. This error would cause the DMA controller to transfer more data than we expected.
In your test environment I think the problem is that with word size, the DMA controller would be trying to access some address outside our src/dst data structures and it maybe illegle, so the program got stuck.

cybertale replied:
According to the Reference Manual (Which I'm referencing now is 9.3.10 of RM0368, STM32F401xB), When the source data size and dest data size is different, the NDT is determined by PSIZE. So we should fix this to  
```suggestion
if (source_periph) {
	DMA_InitStruct.NbData = config->head_block->block_size / config->source_data_size;
} else {
	DMA_InitStruct.NbData = config->head_block->block_size / config->dest_data_size;
}
```

cybertale commented:
# Change Log
1. NDT is now determined by peripheral size. @maxvankessel 
2. Add one commit 558016a to change the block size in the i2s driver from 1 to 2. This should be fixed no matter if the new driver is used. @avisconti can have a test on this, but I think change of this time will not affect behaviour of the i2s driver at all since NDT is still the same value as before after the calculation(before is 1, now is 2/2 also 1).

maxvankessel commented on drivers/dma/dma_stm32.c:
```diff
@@ -0,0 +1,621 @@
+		return -EINVAL;
+	}
+
+	LL_DMA_SetDataLength(dma, table_ll_stream[id], size);
```
You might want to check this one to.

cybertale replied:
Right

cybertale committed:
boards: arm: 96b_argonkey: Update dma defconfig for tests

Previous defconfig for dma is DMA_STM32F4X in this board, while the new
generic driver uses DMA_STM32 to enable DMA support, and also dma driver
of stm32 now needs HEAP_MEM_POOL_SIZE to be big enough to hold dma
stream instances.

Additional .conf files are added for also adding HEAP_MEM_POOL_SIZE
configuration to two test cases.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
drivers: i2s: apply new data size for DMA interface

the old DMA driver used to use 1 to stand for 16 bits, while the
new driver uses 2 to stand for 16 bits, which means the
'source_data_size' and the 'dest_data_size' should be 2.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
drivers: i2s: fix block size for DMA interface

`block_count` in `dma_cfg` is described as how many bytes to be
transfered in dma.h. So it should be 2 since the source data size
and dest data size are all 16 bits in this application. And all
block size should represent just bytes.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
boards: 96b_stm32_sensor_mez: apply kconfig changes

DMA is now selected by Kconfig symbols in I2S, and DMA_STM32
is selected by DMA in
soc/arm/st_stm32/common/Kconfig.defconfig.series. So remove the
DMA selecting operation here.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale commented:
@avisconti Please have a look on the change I just made in f03e093 , have a test to see if the driver is still working.

avisconti commented:
> @avisconti Please have a look on the change I just made in [f03e093] have a test to see if the driver is still working.

@cybertale 
Let me do it on Monday. Today I'm busy with other stuff...

cybertale commented:
> > @avisconti Please have a look on the change I just made in [[f03e093](https://github.com/zephyrproject-rtos/zephyr/commit/f03e09353c3304d02d05334e319b95eaccac4a54)] have a test to see if the driver is still working.
> 
> @cybertale
> Let me do it on Monday. Today I'm busy with other stuff...

Sure.

avisconti commented:
> > > @avisconti Please have a look on the change I just made in [[f03e093](https://github.com/zephyrproject-rtos/zephyr/commit/f03e09353c3304d02d05334e319b95eaccac4a54)] have a test to see if the driver is still working.
> > 
> > 
> > @cybertale
> > Let me do it on Monday. Today I'm busy with other stuff...
> 
> Sure.

@cybertale @erwango 
I re-tested it and it works fine on ArgonKey board using the microphone sample.

erwango commented:
^^ @galak

galak commented:
Can you rebase this to deal with the merge conflict?

galak commented:
In general this looks good, made some minor changes to the DTS bindings.  Also pulled in a fix from 20362.  I'd like to hold or pull out the new test into its own PR.

cybertale commented:
@galak Do you mean I should move some commits from this PR to another PR?

galak commented:
> @galak Do you mean I should move some commits from this PR to another PR?

can you open a PR with just the 'boards: arm: add dma test support for some nucleo board'.

galak commented:
for the I2S driver can we not determine CLK_SEL_n from DTS?

cybertale commented:
@galak Sorry, just got the notification email. I'll create another PR for the nucleo boards, thanks!

cybertale commented:
> 
> 
> for the I2S driver can we not determine CLK_SEL_n from DTS?

@avisconti

avisconti commented:
> > for the I2S driver can we not determine CLK_SEL_n from DTS?
> 
> @avisconti

I'll take a look later.
Maybe @erwango has something to say about it.

erwango commented:
@cybertale, any news of boards: arm: add dma test support for some nucleo board' ?
